filestream/filestream.go

```diff
@@ -0,0 +1,106 @@
+package filestream
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/fsnotify/fsnotify"
+)
+
+type FileStream struct {
+	file_name string
+	reader    *os.File
+	watcher   *fsnotify.Watcher
+
+	// Used to abort streaming when a client disconnects, etc
+	ctx context.Context
+}
+
+//-------------------------------------------------------------------------------------------------
+func New(ctx context.Context, file_name string) (*FileStream, error) {
+	file, err := os.Open(file_name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open file '%s': %w", file_name, err)
+	}
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, fmt.Errorf("failed to start a watcher: %w", err)
+	}
+
+	err = watcher.Add(file_name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to add file '%s' to the watcher: %w", file_name, err)
+	}
+
+	return &FileStream{
+		file_name: file_name,
+		reader:    file,
+		watcher:   watcher,
+		ctx:       ctx,
+	}, nil
+}
+
+//-------------------------------------------------------------------------------------------------
+// Closes the fsnotify watcher and the underlying file stream
+func (s *FileStream) Close() error {
+	err := s.watcher.Close()
+	if err == nil {
+		err = s.reader.Close()
+	}
+	return err
+}
+
+//-------------------------------------------------------------------------------------------------
+func (s *FileStream) WaitForChanges() (changed bool) {
+	result := make(chan bool)
+	go func() {
```

Comment:
If this goroutine is truly needed due to limitations in `fsnotify`, perhaps `FileStream` could own the `result` chan and just launch one goroutine that monitors the events in `New`. Then in `Read` you could just block on reading from the channel instead. Thoughts?

Reply:
I've tried it and it looked pretty confusing to me, so I opted for a pattern where the concurrency is contained within a single function and it is much easier to reason about IMO.

Reply:
Actually, I have refactored the code slightly and looks like the goroutine is not needed in there (at least I was not able to find any issues with not using it)